### ISSUE TITLE
Pixiv disabled blocking of API requests via cloudflare

### ISSUE THREAD

**Issue by theBreadCompany** [upbit/pixivpy#171]
I recently noticed that its possible to use the pixiv.net APIs without cloudscraper. This might be kind of irrelevant for Python as the cloudscraper module isn't that big, but it isn't for any other languages like Swift, which don't have modules to workaround cloudflare. I really hope that this is not just a mistake by Pixiv...
Just fyi...

**Comment by grawity** [comment]
The main APIs indeed seem to be available without Cloudflare, but unfortunately I still keep getting a Cloudflare block _almost every time_ when renewing the OAuth2 token.

> requests POST https\://oauth.secure.pixiv.net/auth/token error: Detected a Cloudflare version 2 Captcha challenge, This feature is not available in the opensource (free) version.

But it seems that _python-requests, specifically,_ is getting blocked by Cloudflare while other HTTP clients are not (e.g. python-httpx or pycurl have no problems).

The important difference seems to be in the TLS handshake that python-requests (or rather urllib3) performs, not in the request headers. Python-requests gets blocked because it sets OP_NO_TICKET and doesn't send the "session_ticket" extension. Repeating exactly the same request, byte-to-byte, but with this option removed isn't blocked.

My current workaround is:

``` python
class CustomAdapter(requests.adapters.HTTPAdapter):
    def init_poolmanager(self, *args, **kwargs):
        # When urllib3 hand-rolls a SSLContext, it sets 'options |= OP_NO_TICKET'
        # and CloudFlare really does not like this. We cannot control this behavior
        # in urllib3, but we can just pass our own standard context instead.
        import ssl
        ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
        ctx.load_default_certs()
        ctx.set_alpn_protocols(["http/1.1"])
        return super().init_poolmanager(*args, **kwargs, ssl_context=ctx)

...
api.requests = requests.Session()
api.requests.mount("https://", CustomAdapter())
```

Possible other workaround: Completely replace python-requests with python-httpx (which is API-compatible).

``` python
api.requests = httpx.Client()
```

**Comment by grawity** [comment]
Oh, 🤦  I just realized that this is related to the combination of User-Agent _and_ TLS handshake. I guess Cloudflare might be matching the TLS options against what the real "PixivAndroidApp" would normally use, because if I remove the custom User-Agent and just use the standard python-requests User-Agent, I'm no longer blocked as well.

I guess it makes a lot of sense for a bot-protection system to do that...

So I'm not sure what's better for pixivpy – fixing the ssl_context to better imitate PixivApp, or fully reverting to default python-requests behavior? (It seems I can actually refresh tokens _without_ sending any User-Agent or X-Client headers...)

**Comment by upbit** [comment]
If Cloudflare only judges based on `User-Agent`, may be we can provide additional parameters to disguise auth:

```py
    # Usage: api.auth(refresh_token="xxx", ext_headers={"User-Agent": "python-requests/2.18.4"})
    def auth(self, username=None, password=None, refresh_token=None, ext_headers={}):
        headers = {
            'User-Agent': 'PixivAndroidApp/5.0.115 (Android 6.0; PixivBot)',
            'X-Client-Time': local_time,
            'X-Client-Hash': hashlib.md5((local_time + self.hash_secret).encode('utf-8')).hexdigest(),
        }
        for key, val in ext_headers.items():
            headers[key] = val
```

**Comment by theBreadCompany** [comment]
Sorry for not commenting for several days. I can confirm that simply reverting to python-requests (replacing cloud scraper.create_scraper() with requests in line 31 of api.py) solves the blocks. Thank you very much for your work.

**Comment by upbit** [comment]
Fixed in v3.6.0:

1. Modified the default `User-Agent` to the latest iOS version number.
2. If you want to modify it into your own UA, you can write like this:

`api.auth(refresh_token=_REFRESH_TOKEN, headers={"user-agent": "PixivIOSApp/7.13.3 (iOS 15.0; iPhone13,4)"})`
